staking/programs/staking/src/state/positions.rs

```diff
@@ -3,59 +3,152 @@ use {
+            }
+        }
+
+        Ok(Self { acc_info })
+    }
+
+    pub fn load_mut(account_loader: &AccountLoader<'a, PositionData>) -> Result<Self> {
+        let result = Self::load(account_loader)?;
+        if !result.acc_info.is_writable {
+            return Err(AnchorErrorCode::AccountNotMutable.into());
+        }
+        Ok(result)
+    }
+
+    pub fn load(account_loader: &AccountLoader<'a, PositionData>) -> Result<Self> {
+        let acc_info = account_loader.to_account_info();
+
+        {
+            let data = acc_info.try_borrow_data()?;
+            if data.len() < PositionData::discriminator().len() {
+                return Err(AnchorErrorCode::AccountDiscriminatorNotFound.into());
+            }
+
+            let disc_bytes = array_ref![data, 0, 8];
+            if disc_bytes != &PositionData::discriminator() {
+                return Err(AnchorErrorCode::AccountDiscriminatorMismatch.into());
+            }
+        }
+        Ok(Self { acc_info })
+    }
+
+    pub fn add_rent_if_needed(&self, payer: &Signer<'a>) -> Result<()> {
+        let rent = Rent::get()?;
+        let amount_to_transfer = rent
+            .minimum_balance(self.data_len())
+            .saturating_sub(self.acc_info.lamports());
+
+        if amount_to_transfer > 0 {
+            let transfer_instruction =
+                system_instruction::transfer(payer.key, self.acc_info.key, amount_to_transfer);
+
+            anchor_lang::solana_program::program::invoke(
+                &transfer_instruction,
+                &[payer.to_account_info(), self.acc_info.clone()],
+            )?;
+        }
+        Ok(())
+    }
+
+    pub fn owner(&self) -> Result<Pubkey> {
+        let data = self.acc_info.try_borrow_data()?;
+        Ok(Pubkey::from(*array_ref![data, 8, 32]))
+    }
+
+    pub fn set_owner(&self, owner: &Pubkey) -> Result<()> {
+        let mut data = self.acc_info.try_borrow_mut_data()?;
+        data[8..40].copy_from_slice(&owner.to_bytes());
+        Ok(())
+    }
+
+    pub fn get_position_capacity(&self) -> usize {
+        self.acc_info.data_len().saturating_sub(PositionData::LEN) / POSITION_BUFFER_SIZE
     }
 
     /// Finds first index available for a new position, increments the internal counter
     pub fn reserve_new_index(&mut self, next_index: &mut u8) -> Result<usize> {
-        let res = *next_index as usize;
-        *next_index += 1;
-        if res < MAX_POSITIONS {
-            Ok(res)
-        } else {
-            Err(error!(ErrorCode::TooManyPositions))
+        let position_capacity: usize = self.get_position_capacity();
+        let res = usize::from(*next_index);
+        *next_index = next_index
+            .checked_add(1)
+            .ok_or_else(|| error!(ErrorCode::TooManyPositions))?;
+
+        if res == position_capacity {
+            self.acc_info.realloc(
+                PositionData::LEN + usize::from(*next_index) * POSITION_BUFFER_SIZE,
+                false,
+            )?;
```

Comment:
What happens if the current len is larger? do we reduce size?

Reply:
Future PR